src/target/riscv/riscv.c

```diff
@@ -2712,6 +2712,13 @@ static int riscv_get_gdb_reg_list(struct target *target,
 
 static int riscv_arch_state(struct target *target)
 {
+	assert(target->state == TARGET_HALTED);
+	LOG_USER("%s halted due to %s.%s",
+			target_name(target),
+			debug_reason_name(target),
+			(target->semihosting && target->semihosting->is_active) ?
+			" Semihosting is active." :
+			"");
 	struct target_type *tt = get_target_type(target);
 	if (!tt)
 		return ERROR_FAIL;
```

Comment:
Another optional detail: Since the function pointers in `riscv_target`, `riscv011_target`, `riscv013_target` and in `arch_info` (and maybe others) are becoming not easy to orient in, I'd use the opportuninty and add here an assert for additional safety. 

```suggestion
		return ERROR_FAIL;
	assert(tt->arch_state);
```

Reply:
Addressed.